csharp/src/Ice/Communicator.cs

```diff
@@ -1011,6 +1016,20 @@ async Task PerformDisposeAsync()
         public IEndpointFactory? IceFindEndpointFactory(Transport transport) =>
             _transportToEndpointFactory.TryGetValue(transport, out IEndpointFactory? factory) ? factory : null;
 
+        /// <summary>Add one or more dispatch interceptors to the list of dispatch interceptors, the dispatch
+        /// interceptors registered with a communicator will run after the interceptors registered with an object
+        /// adapter, and in the order that they are registered.</summary>
+        /// <param name="interceptors">The dispatch interceptors to register with the communicator.</param>
+        public void InterceptDispatch(params DispatchInterceptor[] interceptors) =>
+            _dispatchInterceptors.AddRange(interceptors);
+
+        /// <summary>Add one or more invocation interceptors to the list of invocation interceptors, the invocation
+        /// interceptors registered with the communicator are executed in the order they are registered for each
+        /// invocation done with the communicator.</summary>
+        /// <param name="interceptors">The invocation interceptors to register with the communicator.</param>
+        public void InterceptInvocation(params InvocationInterceptor[] interceptors) =>
+            _invocationInterceptors.AddRange(interceptors);
+
```

Comment:
Is this thread safe?

Reply:
I would add the interceptor registration to the Communicator ctor to avoid this issue.

Reply:
It is not, we can use a thread safe collection.

Reply:
Locking this list on every invocation doesn't seem great either. Maybe they should be added during communicator consturciton?

Reply:
I think will be good that plug-ins can add interceptors, but for this to work they have to call a method on the communicator.